core/payment/src/api/allocations.rs

```diff
@@ -32,187 +33,252 @@ const DEFAULT_TESTNET_NETWORK: NetworkName = NetworkName::Holesky;
+                                "Network not specified, using default {}",
+                                DEFAULT_TESTNET_NETWORK
+                            );
+                            DEFAULT_TESTNET_NETWORK.into()
+                        }
+                    } else {
+                        log::debug!(
+                            "Network not specified and token not specified, using default {}",
+                            DEFAULT_TESTNET_NETWORK
+                        );
+                        DEFAULT_TESTNET_NETWORK.into()
+                    }
+                });
+                let network = validate_network(network_str).map_err(|err| {
+                    bad_req_and_log(format!("Validate network failed (1): {err}"))
+                })?;
+
+                let driver_str = p.driver.as_deref().unwrap_or_else(|| {
                     log::debug!(
-                        "Network not specified, using default {}, because token set to glm",
-                        DEFAULT_MAINNET_NETWORK
+                        "Driver not specified, using default {}",
+                        DEFAULT_PAYMENT_DRIVER
                     );
-                    DEFAULT_MAINNET_NETWORK.into()
+                    DEFAULT_PAYMENT_DRIVER.into()
+                });
+                let driver = validate_driver(&network, driver_str)
+                    .map_err(|err| bad_req_and_log(format!("Validate driver failed (1): {err}")))?;
+
+                if let Some(token) = p.token.as_ref() {
+                    let token =
+                        TokenName::from_token_string(&driver, &network, token).map_err(|err| {
+                            bad_req_and_log(format!("Validate token failed (1): {err}"))
+                        })?;
+                    log::debug!("Selected network {}-{}-{}", driver, network, token);
+                    Self {
+                        driver,
+                        network,
+                        token,
+                    }
                 } else {
+                    let default_token = TokenName::default(&driver, &network);
+
                     log::debug!(
-                        "Network not specified, using default {}",
-                        DEFAULT_TESTNET_NETWORK
+                        "Selected network with default token {}-{}-{}",
+                        driver,
+                        network,
+                        default_token
                     );
-                    DEFAULT_TESTNET_NETWORK.into()
+                    Self {
+                        driver,
+                        network,
+                        token: default_token,
+                    }
                 }
-            } else {
-                log::debug!(
-                    "Network not specified and token not specified, using default {}",
-                    DEFAULT_TESTNET_NETWORK
-                );
-                DEFAULT_TESTNET_NETWORK.into()
-            }
```

Comment:
This could be match statement probably:
```
match (p.driver, p.network, p.token) {
    (None, None, None) => { ... },
    (None, None, Some(token)) => { ... }
}
```